docs/markdown/Python/python-integrations/google-cloud-function-python.md

```diff
@@ -59,32 +69,56 @@ Pants will use [dependency inference](doc:targets) based on the `handler` field,
-  Runtime: python3.8
+$ pants package project/:cloud_function
+Wrote dist/project/cloud_function.zip
   Handler: handler
 ```
 
 > 🚧 Running from macOS and failing to build?
-> 
+>
 > Cloud Functions must run on Linux, so Pants tells PEX and Pip to build for Linux when resolving your third party dependencies. This means that you can only use pre-built [wheels](https://packaging.python.org/glossary/#term-wheel) (bdists). If your project requires any source distributions ([sdists](https://packaging.python.org/glossary/#term-source-distribution-or-sdist)) that must be built locally, PEX and pip will fail to run.
-> 
+>
 > If this happens, you must either change your dependencies to only use dependencies with pre-built [wheels](https://pythonwheels.com) or find a Linux environment to run `pants package`.
 
 Step 4: Upload to Google Cloud
 ------------------------------
 
 You can use any of the various Google Cloud methods to upload your zip file, such as the Google Cloud console or the [Google Cloud CLI](https://cloud.google.com/functions/docs/deploying/filesystem#deploy_using_the_gcloud_tool).
 
-You must specify the handler as `handler`.
+You must specify the handler as `handler`. This is a re-export of the function referred to by the `handler` field of the target.
+
+Migrating from Pants 2.16 and earlier
+-------------------------------------
+
+Pants has implemented a new way to package Google Cloud Functions in 2.17, resulting in smaller packages and faster cold starts. This involves some changes:
+
+- In Pants 2.16 and earlier, Pants used the [Lambdex](https://github.com/pantsbuild/lambdex) project. First, Pants would convert your code into a [Pex file](doc:pex-files) and then use Lambdex to adapt this into a zip file understood by GCF by adding a shim handler. This shim handler first triggers the Pex initialization to choose and unzip dependencies, during initialization.
+- In Pants 2.17, the use of Lambdex is deprecated, in favour of choosing the appropriate dependencies ahead of time, as described above, without needing to do this on each cold start. This results in a zip file laid out in the format recommended by GCF, and includes a re-export of the handler.
+- In Pants 2.18, the new behaviour will become the default behaviour.
+- In Pants 2.19, the old Lambdex behaviour will be entirely removed.
+
+Any existing `python_google_cloud_function` targets will change how they are built. Migrating has three steps:
+
+1. opt-in to the new behaviour in Pants 2.17
+2. package the new targets
+3. upload those packages to GCF (the existing handler configuration should still work)
+
+To opt-in to the new behaviour in Pants 2.17, set:
+
+``` toml pants.toml
+[lambdex]
+layout = "zip"
+```
+
+To temporarily continue using the old behaviour in Pants 2.17, instead set `layout = "lambdex"`. This will not be supported in Pants 2.19. If you encounter a bug with `layout = "zip"`, [please let us know](https://github.com/pantsbuild/pants/issues/new/choose).
```

Comment:
Theoretically, we could also document the ability to use a `pex_binary` as a Lambda artefact directly (e..g if a user needs the dynamic initialisation), with the `__pex__.` prefix on the handler. Thoughts?

Reply:
Can't hurt! But it is an advanced topic, and should be introduced as such. "In the rare case where you need to X..."

Reply:
I've added a dedicated section "Advanced: Using PEX directly" section, outside this one (but linked to it), because those docs seem like they'll be relevant even after the transition is ancient history.